PV resizing logic

pranavgaikwad opened the pull request:
Signed-off-by: Pranav Gaikwad <pgaikwad@redhat.com>

This PR: 
- Adds DF command basics
- Adds ResticDFExecutor, responsible for running df using Restic pods
- Adds VolumeAdjuster responsible for computing adjusted volume size

shawn-hurley commented on pkg/controller/miganalytic/miganalytics_controller.go:
```diff
@@ -108,9 +108,9 @@ type ReconcileMigAnalytic struct {
 type MigAnalyticPersistentVolumeDetails struct {
 	Name                string
-	RequestedCapacity   *resource.Quantity
+	RequestedCapacity   resource.Quantity
```
Sorry, if this is a small change, but is there a distinction between no RequestedCapacity and `0` RequestedCapacity? Could there already be a stored version on a cluster with nil here and would that cause issues on upgrade?

pranavgaikwad replied:
@shawn-hurley valid question. AFAIK, that shouldn't cause the problem, the new struct will default to `0` values. That case will be handled further in the VolumeAdjustment logic

shawn-hurley replied:
Wouldn't you have a situation like this? because we don't have omit-empty?

https://play.golang.org/p/RQQFv9_mfS-

pranavgaikwad replied:
@shawn-hurley I see your point now. This is a new change we introduced specifically for PV resize stuff in an earlier API PR. So we don't have this struct present on any MigAnalytics in previous versions. MigAnalytic controller fills this when it finds the PVs. So I think we wont come across scenario you described

shawn-hurley commented on pkg/controller/miganalytic/restic_executor.go:
```diff
@@ -0,0 +1,150 @@
+	for node := range pvcNodeMap {
+		resticPodRef := r.getResticPodForNode(node)
+		// if no Restic pod is found for this node, all PVCs on this node are skipped
+		if resticPodRef == nil {
```
if we hit this condition, we won't make on of the wait-groups as done and we will look I think.

pranavgaikwad replied:
@shawn-hurley good catch!

shawn-hurley commented on pkg/controller/miganalytic/restic_executor.go:
```diff
@@ -0,0 +1,150 @@
+			}
+			continue
+		}
+		go func(n string, podRef *corev1.Pod, outputs map[string]DFCommand) {
```
it does not appear that there is any limit or buffer, so lets say we have 500 nodes we will have 500 threads, and then we will quickly make 500 calls and then they will block until the map is updated. 

Not only will this not really gain us as much value from the async (I think right?) it will spam the API server will a ton of requests really quickly. Assuming that each call takes some period of time, we may want to consider making a buffered channel executor that also parses the output from below in each thread and the return is pvcDFInfo?

WDYT? too much over engineering?

pranavgaikwad replied:
@shawn-hurley good point. I think it's fair to assume a scenario for 500 nodes. I can use your help implementing some sort of a buffered channel. Haven't really done that.

alaypatel07 replied:
+1 good idea

shawn-hurley commented on pkg/controller/miganalytic/volume_adjustment.go:
```diff
@@ -0,0 +1,249 @@
+
+// GetDFOutputForPV given a volume name and pod uid, returns structured df ouput for the volume
+// only works on outputs of commands created by DFCommand.PrepareDFCommand()
+func (cmd *DFCommand) GetDFOutputForPV(volName string, podUID types.UID) (pv DFOutput) {
```
I probably would like a unit test for this, I feel like this could be pretty brittle as we are parsing console out

Have you considered using some `awk` to format the output as JSON?

pranavgaikwad replied:
@shawn-hurley since we are running DF for all volumes on a node at once, I thought it might be just easier to parse the output in golang. For instance, we will generate command like this:

```sh
df -BM /path-1 /path-2 /path-3 /path-4 /path-5 ...
```

The logic is very simple. We just look for volume names in output lines (StdErr/StdOut).

shawn-hurley replied:
The logic may be simple, but what if we get back something we don't know how to handle? 

I think why I think a unit test is important, is AFAICT we run a bunch of regex's to find the info, let's make sure that if there is a bug in here and someone tries to fix it, that they don't regress on the output that we are expecting.

The reason for JSON or some other format is probably not needed as I doubt DF will be changing its output structure.

pranavgaikwad replied:
@shawn-hurley sounds good. Actually there's already a unit test for this function. Looking at it, I realize that I should add more cases to it which handle unparseable inputs.

djwhatle replied:
+1 this seems like a great place for unit tests, pretty self contained input/output

pranavgaikwad replied:
updated test

shawn-hurley commented on pkg/controller/miganalytic/volume_adjustment.go:
```diff
@@ -0,0 +1,249 @@
+		Status: True,
+		Type: ExtendedPVAnalysisFailed,
+		Reason: FailedRunningDf,
+		Message: fmt.Sprintf(warningStrings, pvNames),
```
s/warningStrings/waringString?

shawn-hurley commented on pkg/controller/miganalytic/volume_adjustment.go:
```diff
@@ -0,0 +1,249 @@
+	pvNames := ""
+	for _, dfOutput := range erroredPVs {
+		if dfOutput.IsError {
+			fmt.Sprintf(
```
is this supposed to saved as pvNames?

shawn-hurley commented on pkg/controller/miganalytic/volume_adjustment.go:
```diff
@@ -0,0 +1,249 @@
+}
+
+// generateWarningForErroredPVs given a list of dfoutputs, generates warnings for associated pvcs
+func (pva *PersistentVolumeAdjuster) generateWarningForErroredPVs(erroredPVs []*DFOutput) {
```
A unit test here might a good thing as well?

shawn-hurley commented on pkg/controller/miganalytic/volume_adjustment.go:
```diff
@@ -0,0 +1,249 @@
+
+func (pva *PersistentVolumeAdjuster) calculateProposedVolumeSize(usagePercentage int64, actualCapacity resource.Quantity,
+	requestedCapacity resource.Quantity) (proposedSize resource.Quantity, reason string) {
+	return
```
Assume this still needs to be implemented?

pranavgaikwad replied:
@shawn-hurley correct. Shubham is implementing this piece. We will get this in a different PR

shawn-hurley commented on pkg/controller/miganalytic/volume_adjustment_test.go:
```diff
@@ -0,0 +1,273 @@
+	},
+}
+
+func TestDFCommand_convertDFQuantityToKubernetesResource(t *testing.T) {
```
I would like to see a test for GB as well

shawn-hurley commented on pkg/controller/miganalytic/restic_executor.go:
```diff
@@ -0,0 +1,183 @@
+	}
+	err := podCommand.Run()
+	if err != nil {
+		log.Info(
```
how does the user know that we are unsuccessful in determining the current pod volumes, and why

Can't this also error for other reasons like pod does not exist or unable to connect to pod or some other reason?

pranavgaikwad replied:
@shawn-hurley according to current implementation, if this happens, all PVCs on this node will be added to Warning indicating that we failed to gather data for these volumes. It's either we succeed or we fail. Maybe we can log the message here, but unsure how to bubble up the detailed failure information all the way up until MigPlan.

pranavgaikwad commented on pkg/controller/miganalytic/restic_executor.go:
```diff
@@ -0,0 +1,156 @@
+			output := r.DF(podRef, pvcNodeMap[n])
+			mutex.Lock()
+			dfOutputs[n] = output
+			mutex.Unlock()
```
Use `defer`

djwhatle reviewed (commented):
I think we need to be sure that this triggers at the right times, and what we're lacking right now is a trigger on namespace change. Pranav and I discussed adding a `status.NamespaceDigest` field to MigPlan and also a `HasReconciledNamespaces()` function to help  with this. 

Down the road (or if Pranav feels up to changing this now ;) )we can also use this field to limit PV discovery to only run when 
1. MigPlan is new
2. Namespaces change 
3. Refresh is requested.

Also think we should shrink the footprint of the PV resize logic in the main reconcile func and move some of that code out to a func called from main reconcile. 

For example of deciding whether to trigger an action based on `HasReconciled()`: https://github.com/konveyor/mig-controller/blob/master/pkg/controller/migplan/gvk.go#L12-L16

Edit: There are several open PRs related to PV intelligent resize, this comment is generally applicable to these PRs but perhaps not relevant to content of this particular PR.

pranavgaikwad commented:
I have addressed all concerns folks had on the contents of this PR. There is one open concern raised by @shawn-hurley about error handling. Since 1.4.1 is just API support for the feature, I'd like to merge this for now without addressing error handling and come up with a holistic solution for that case when we start working on UI/UX for the feature. In effect, this PR alone is just focuses on creating a MigAnalytic flow for collecting DF information from Restic pods, and doesn't affect / break any existing features.

alaypatel07 commented on pkg/controller/miganalytic/miganalytics_controller.go:
```diff
@@ -236,8 +236,10 @@ func (r *ReconcileMigAnalytic) analyze(analytic *migapi.MigAnalytic) error {
+	nodeToPVMap := make(map[string][]MigAnalyticPersistentVolumeDetails)
 
+	analytic.Status.Analytics.Plan = plan.Name
+	analytic.Status.Analytics.Namespaces = make([]migapi.MigAnalyticNamespace, 0)
```
do we know the namespaces length already? can we seen it with length that is not 0?

alaypatel07 replied:
FTR its a NIT, not a merge blocker

djwhatle commented on pkg/controller/miganalytic/restic_executor.go:
```diff
@@ -0,0 +1,156 @@
+		waitGroup.Add(1)
+		go func(n string, podRef *corev1.Pod) {
+			// block until channel empty
+			bufferedExecutionChannel <- struct{}{}
```
Worth noting for any other reviewers looking at this scratching their heads, this is just stuffing a blank struct in the channel as a means of messaging. No actual data is contained.

alaypatel07 commented:
glanced quickly, love the unit tests and buffered reader implementation, +1, will leave the ack for other folks who have followed more closely, but lgtm

djwhatle commented on pkg/controller/miganalytic/restic_executor.go:
```diff
@@ -0,0 +1,156 @@
+}
+
+// Execute given a map node->[]pvc, runs Df command for each, returns list of structured df output per pvc
+func (r *ResticDFCommandExecutor) Execute(pvcNodeMap map[string][]MigAnalyticPersistentVolumeDetails) ([]DFOutput, error) {
```
I wonder if we should consider bumping `MaxConcurrentReconciles` to `10` instead of `1` so that the PV resize MigAnalytic is never stuck waiting in line behind the general purpose MigAnalytic that pulls all resource counts for plan.

https://github.com/konveyor/mig-controller/blob/master/pkg/controller/miganalytic/miganalytics_controller.go#L79
